src/sdk/python/src/iotea/core/protocol_gateway.py

```
@@ -0,0 +1,235 @@
+import asyncio
+import importlib
+import json
+
+from .util.json_model import JsonModel
+
+class ProtocolGateway:
+    def __init__(self, protocol_gateway_config, display_name, use_platform_protocol_only=False):
+        ProtocolGateway.validate_configuration(protocol_gateway_config, use_platform_protocol_only)
+        pg_config = JsonModel(protocol_gateway_config)
+        self.use_platform_protocol_only = use_platform_protocol_only
+        self.adapters = []
+
+        for adapter_config in pg_config.get('adapters'):
+            adapter_config_model = JsonModel(adapter_config)
+            is_platform_protocol = adapter_config_model.get('platform', False)
+
+            if self.use_platform_protocol_only and not is_platform_protocol:
+                # Skip all non-platform protocols
+                continue
+            module = importlib.import_module(adapter_config_model.get('module.name'), package=__package__)
+            a_class = getattr(module, adapter_config_model.get('module.class'))
+            instance = a_class(adapter_config_model.get('config'), display_name)
+
+            self.adapters.append(Adapter(instance, is_platform_protocol))
+
+
+    async def publish(self, topic, message, publish_options=None, force_wait=False):
+        if publish_options is None:
+            publish_options = ProtocolGateway.create_publish_options()
+
+        publish_to_platform_protocol_only = publish_options.platform_protocol_only
+
+        if publish_to_platform_protocol_only is None:
+            # Apply global settings, if nothing is given
+            publish_to_platform_protocol_only = self.use_platform_protocol_only
+
+        self.__validate_platform_protocol_usage(publish_to_platform_protocol_only)
+
+        for adapter in self.adapters:
+            if publish_to_platform_protocol_only is False or adapter.is_platform_protocol:
+                if publish_options.adapter_id is None or publish_options.adapter_id is adapter.id:
```

Using `is` and `is not` is meant for comparing object identity. This works for the types `None`, `True` and `False` because the are "interned" (see `sys.intern`) but generally does not work when comparing object values for equality. In this case the comparison `publish_options.adapter_id is adapter.id` may fail unexpectedly. Please change it to `publish_options.adapter_id == adapter.id`